### PR TITLE
Fix memory leaks

### LINKED ISSUE #1047: Fix memory leaks in bpftrace
If you run a trivial bpftrace script with ASAN on (see #1046), you'll see ~34 memory leaks: http://ix.io/2628 .

Build with asan by:
```
$ cd ~/dev/bpftrace
$ mkdir build && cd build
$ cmake .. -DBUILD_ASAN && make -j4
$ sudo ./src/bpftrace -e 'BEGIN { printf("hello\n"); }'
[...]
```

I'm sure we can find a lot more by running some of the scripts in `tools/` under asan mode.

### PR DESCRIPTION
**Pull request by danobi**
This PR fixes some memory leaks in bpftrace. I've only run the asan
build under some very basic programs so this doesn't prove that bpftrace
is totally leak free or anything.

This PR is also rebased on top of #1050. I'll rebase this PR onto master 
once #1050 merges.

This fixes some of #1047.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -490,10 +490,6 @@ int main(int argc, char *argv[])
   if (!clang.parse(driver.root_, bpftrace, extra_flags))
     return 1;
 
-  err = driver.parse();
```
Comment (ajor) [inline src/main.cpp]: I think this second parse is needed to handle `#defines` (but a comment explaining that would be good if I'm right)

**Comment by danobi** [comment]
@ajor I tried that initially (I mention it in one of the commit messages) but I didn't really like how that turned out. Here's the code: https://github.com/danobi/bpftrace/commit/511d1927ac60b5d05a3a8bb7fa3a954c7586f147

It seemed like a huge amount of churn in parser.yy. What are your thoughts?

**Comment by danobi** [comment]
So bison recently (see "C++: Support for move semantics (lalr1.cc)" in https://fossies.org/linux/bison/NEWS) added support for smart pointers. But unfortunately bison 3.2.1 (2018-11-09) isn't widely distributed yet. I'm running fedora 30 and still on bison-3.0.5.

I'm trying to do build time patching of `parser.tab.hh` but it's really heinous and fragile. I think the most maintainable solution is to do the regular old `delete`s for now.

**Comment by ajor** [comment]
Sorry, I meant to look into this to try and understand but didn't get round to it. Is it just that we have to explicitly type out `std::move()` or do move semantics not work at all with older Bisons? What's broken with them?

**Comment by danobi** [comment]
It's more than typing out std::move, it's that bison is missing an std::move in the generated code: http://www.comrite.com/wp/c11-stdunique_ptr-as-semantic-value-in-bison-c-mode/

**Comment by danobi** [comment]
There's also some issues with ownership in ast.h. For example, `@a <<= 1` creates an `ast::AssignMapStatment` and an `ast::Binop`, both of which own pointers to the underlying `ast::Map`.

Even old style `delete`s get kind of nasty here b/c you can't always assume the AssignMapStatement holds ownership.

Could go with shared_ptr but that just kicks the ownership issues down the road.

**Comment by danobi** [comment]
I pushed some code that gets the job done. Not very pretty though.

**Comment by ajor** [comment]
Ah yeah we'd have to change the compound assignments. I *think* something like this should work to make a copy of one of the `$1`s to fix that problem:
```
$$ = make_unique<ast::AssignMapStatement>(
       $1,
       make_unique<ast::Binop>(std::make_unique<Map>(*$1),
                               token::LEFT,
                               $3,
                               @2),
       true);
```

**Comment by ajor** [comment]
Ok, I guess raw pointers are the way to go. We should also delete the default copy and move constructors to complete the "rule of 5".

I'd actually like to add copy constructors for these nodes at some point so we can do some AST manipulation, but that can wait. When we add copy constructors we'll be able to change the compound assignment node to have two separate copies of the map and simplify the ownership logic.

BTW, when I was talking about liking inline function definitions I only meant absolutely tiny ones (0/1 lines) ;)   I agree that bigger ones like this make it harder to read

**Pull request closed without merging**